src/libbson/src/bson/bson-atomic.h

```diff
@@ -25,85 +25,283 @@
+         DEF_ATOMIC_CMPEXCH (VCIntrinSuffix,                                  \
+                             MSVC_MEMORDER_SUFFIX (_nf),                      \
+                             __ATOMIC_RELAXED,                                \
+                             a,                                               \
+                             actual,                                          \
+                             new_value);                                      \
+         break;                                                               \
+      default:                                                                \
+         BSON_UNREACHABLE ("Invalid bson_atomic_memorder value");             \
+      }                                                                       \
+      return actual;                                                          \
+   }
+
+#define DECL_ATOMIC_STDINT(Name, VCSuffix) \
+   DECL_ATOMIC_INTEGRAL (Name, Name##_t, VCSuffix)
+
+DECL_ATOMIC_STDINT (int8, 8)
+DECL_ATOMIC_STDINT (int16, 16)
+DECL_ATOMIC_STDINT (int32, )
+
+extern int64_t
+_bson_emul_atomic_int64_fetch_add (int64_t volatile *val,
+                                   int64_t v,
+                                   enum bson_atomic_memorder);
+extern int64_t
+_bson_emul_atomic_int64_exchange (int64_t volatile *val,
+                                  int64_t v,
+                                  enum bson_atomic_memorder);
+extern int64_t
+_bson_emul_atomic_int64_compare_exchange (int64_t volatile *val,
+                                          int64_t expect_value,
+                                          int64_t new_value,
+                                          enum bson_atomic_memorder);
+
+extern void
+bson_yield_thread ();
+
+#if !(defined(_M_IX86) || defined(__i686__))
+/* (64-bit intrinsics are only available in x64) */
+DECL_ATOMIC_STDINT (int64, 64)
 #else
-#define __BSON_NEED_BARRIER 1
-#warning "Unknown compiler, using lock for compiler barrier."
-BSON_EXPORT (void)
-bson_memory_barrier (void);
+static BSON_INLINE int64_t
+bson_atomic_int64_fetch_add (int64_t volatile *val,
+                             int64_t v,
+                             enum bson_atomic_memorder order)
+{
+   return _bson_emul_atomic_int64_fetch_add (val, v, order);
+}
+
+static BSON_INLINE int64_t
+bson_atomic_int64_exchange (int64_t volatile *val,
+                            int64_t v,
+                            enum bson_atomic_memorder order)
+{
+   return _bson_emul_atomic_int64_exchange (val, v, order);
+}
+
+static BSON_INLINE int64_t
+bson_atomic_int64_compare_exchange (int64_t volatile *val,
+                                    int64_t expect_value,
+                                    int64_t new_value,
+                                    enum bson_atomic_memorder order)
+{
+   return _bson_emul_atomic_int64_compare_exchange (val, v, order);
+}
 #endif
 
+DECL_ATOMIC_INTEGRAL (int, int, );
```

Comment:
This line appears to be erroring on 32 bit compile tasks: https://spruce.mongodb.com/task/mongo_c_driver_clang38_i386_debug_compile_nosasl_nossl_patch_d8684bed2f5d0763aceb8c03cbce1e2637bc515c_610adb7e2fbabe37ed81f877_21_08_04_18_25_04/logs?execution=0